### PR TITLE
JIT: Adjust physical promotion heuristics

### PR DESCRIPTION
**Pull request by jakobbotsch**
Adjust the heuristics to take into account recent work on liveness and assignment decomposition.  Stop phrasing things in terms of code size (multiplied by basic block weights, which does not make sense).

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS. First round of adjustments, might dial some of this in more later (specifically related to no costing for assignments with overlapped structs, I need to come up with a good way of costing).

[Diffs with physical promotion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=283918&view=ms.vss-build-web.run-extensions-tab)

[Diffs without old promotion](https://dev.azure.com/dnceng-public/public/_build/results?buildId=283919&view=ms.vss-build-web.run-extensions-tab). -6.6 MB on win-x64 over the old heuristics

File: src/coreclr/jit/promotion.cpp [hunk 1]
```diff
@@ -378,52 +385,78 @@ class LocalUses
...
         weight_t costWithout = 0;
 
-        // A normal access without promotion looks like:
-        // mov reg, [reg+offs]
-        // It may also be contained. Overall we are going to cost each use of
-        // an unpromoted local at 6.5 bytes.
-        // TODO-CQ: We can make much better guesses on what will and won't be contained.
-        costWithout += access.CountWtd * 6.5;
+        // We cost any normal access (which is a struct load or store) without promotion at 3 cycles.
+        costWithout += access.CountWtd * 3;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/promotion.cpp:400]: Consider making all these weight factors symbolic so later on we can vary them more readily.

File: src/coreclr/jit/promotion.cpp [hunk 2]
```diff
@@ -378,52 +385,78 @@ class LocalUses
...
+        // * Even if we don't, decomposing a single struct write into many
+        //   field writes is not necessarily profitable (e.g. 16 byte field
+        //   stores vs 1 XMM load/store).
+        //
+        // TODO-CQ: This ends up being a combinatorial optimization problem. We
+        // need to take a more "whole-struct" view here and look at sets of
+        // fields we are promoting together, evaluating all of them at once in
+        // comparison with the covering struct uses. This will also allow us to
+        // give a bonus to promoting remainders that may not have scalar uses
+        // but will allow fully decomposing assignments away.
```
Comment (AndyAyersMS) [inline src/coreclr/jit/promotion.cpp:458]: Sounds ambitious, but I like the idea of costing things as sets.

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/promotion.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/promotion.cpp:400]: Makes sense, I was initially using the perfscore constants here but noticed some oddities (e.g. stack writes are cheaper on arm64 than stack reads but they are the same cost for xarch). I can add some names the next time around.

File: src/coreclr/jit/promotion.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/promotion.cpp:458]: Yeah, hopefully I can approximate this in a simple/cheap way. But I think I will approach this by writing down an integer linear programming form of the problem and see if it gives me any insights.

**Pull request merged**